test/Npgsql.Tests/ReaderTests.cs

```diff
@@ -1113,6 +1113,60 @@ public async Task NullableScalar()
             }
         }
 
+        [Test, IssueLink("https://github.com/npgsql/npgsql/issues/2913")]
+        public async Task ReaderReadingPreviousQueryMessagesBug()
+        {
+            // No point in testing for multiplexing, as every query may use another connection
+            if (IsMultiplexing)
+                return;
+
+            var firstMrs = new ManualResetEventSlim(false);
+            var secondMrs = new ManualResetEventSlim(false);
+
+            var secondQuery = Task.Run(async () =>
+            {
+                firstMrs.Wait();
+                await using var secondConn = await OpenConnectionAsync();
+                using var secondCmd = new NpgsqlCommand(@"SELECT 1; SELECT 2;", secondConn);
+                await using var secondReader = await secondCmd.ExecuteReaderAsync(Behavior | CommandBehavior.CloseConnection);
+
+                // Check, that StatementIndex is equals to default value
+                Assert.That(secondReader.StatementIndex, Is.EqualTo(0));
+                secondMrs.Wait();
+                // Check, that a first query didn't change StatementIndex
+                Assert.That(secondReader.StatementIndex, Is.EqualTo(0));
+            });
+
+            await using (var firstConn = await OpenConnectionAsync())
+            {
+                // Executing a query, which fails with NpgsqlException on reader disposing, as NotExistingTable doesn't exist
+                using var firstCmd = new NpgsqlCommand(@"SELECT 1; SELECT * FROM NotExistingTable;", firstConn);
+                await using var firstReader = await firstCmd.ExecuteReaderAsync(Behavior | CommandBehavior.CloseConnection);
+
+                Assert.That(firstReader.StatementIndex, Is.EqualTo(0));
+
+                firstReader.ReaderClosed += (s, e) =>
+                {
+                    // Starting a second query, which in case of a bug uses firstConn
+                    firstMrs.Set();
+                    // Waiting for the second query to start executing
+                    Thread.Sleep(100);
+                    // After waiting, reader is free to reset prepared statements, which also increments StatementIndex
+                };
+
+                Assert.ThrowsAsync(Is.TypeOf<PostgresException>(), firstReader.NextResultAsync);
```

Comment:
nit: there's generic ThrowsAsync which accepts a lambda, so you don't need Is.TypeOf